File: GLLara/GLLDropTargetView.m
//
//  GLLDropTargetView.m
//  GLLara
//
//  Created by Torsten Kammer on 22.06.17.
//  Copyright © 2017 Torsten Kammer. All rights reserved.
//

#import "GLLDropTargetView.h"

#import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>

#import "GLLDocument.h"
#import "GLLDocumentWindowController.h"
#import "GLLItemDragDestination.h"

@interface GLLDropTargetView()

@property (nonatomic, retain) GLLItemDragDestination *dragDestination;
    
@end

@implementation GLLDropTargetView

- (void)awakeFromNib {
    [self registerForDraggedTypes:@[ UTTypeFileURL.identifier ]];
    self.dragDestination = [[GLLItemDragDestination alloc] init];
}

- (NSDragOperation)draggingEntered:(id<NSDraggingInfo>)sender {
    GLLDocumentWindowController *windowController = self.window.windowController;
    self.dragDestination.document = windowController.document;
    
    return [self.dragDestination itemDraggingEntered:sender];
}

- (BOOL)performDragOperation:(id<NSDraggingInfo>)sender {
    GLLDocumentWindowController *windowController = self.window.windowController;
    self.dragDestination.document = windowController.document;
    
    NSError *error = nil;
    BOOL success = [self.dragDestination performItemDragOperation:sender error:&error];
    if (!success && error)
        [self presentError:error];
    return success;
}

@end

File: GLLara/GLLDropTargetView.swift
//
//  GLLDropTargetView.swift
//  GLLara
//
//  Created by Torsten Kammer on 01.05.24.
//  Copyright © 2024 Torsten Kammer. All rights reserved.
//

import Cocoa

@objc class GLLDropTargetView: NSView {
    private var dragDestination = GLLItemDragDestination()
    
    override func awakeFromNib() {
        registerForDraggedTypes([.fileURL])
    }
    
    override func draggingEntered(_ sender: any NSDraggingInfo) -> NSDragOperation {
        guard let windowController = window?.windowController as? GLLDocumentWindowController, let document = windowController.document as? GLLDocument else {
            return []
        }
        
        dragDestination.document = document;
        
        return dragDestination.itemDraggingEntered(sender)
    }
    
    override func performDragOperation(_ sender: any NSDraggingInfo) -> Bool {
        guard let windowController = window?.windowController as? GLLDocumentWindowController, let document = windowController.document as? GLLDocument else {
            return false
        }
        
        dragDestination.document = document;
        
        do {
            try dragDestination.performItemDragOperation(sender)
            return true
        } catch let error as NSError {
            presentError(error)
            return false
        }
    }
}
